File: DemoApp-ObjC/DemoApp-ObjC/AppLovin/Native Ads/Programmatic/ALDemoNativeAdProgrammaticViewController.m
//
//  ALDemoNativeAdProgrammaticViewController.m
//  iOS-SDK-Demo
//
//  Created by Thomas So on 9/24/15.
//  Copyright © 2015 AppLovin. All rights reserved.
//

#import "ALDemoNativeAdProgrammaticViewController.h"
#import <AppLovinSDK/AppLovinSDK.h>

@interface ALDemoNativeAdProgrammaticViewController()<ALNativeAdLoadDelegate, ALNativeAdPrecacheDelegate, ALPostbackDelegate>
@property (nonatomic, strong) ALNativeAd *cachedNativeAd;
@property (nonatomic, strong) ALNativeAd *nativeAd;
@end

// Additional documentation - https://applovin.com/integration#iosNative

@implementation ALDemoNativeAdProgrammaticViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.appIcon.layer.masksToBounds = YES;
    self.appIcon.layer.cornerRadius = 3.0f;
    
    self.ctaButton.layer.masksToBounds = YES;
    self.ctaButton.layer.cornerRadius = 3.0f;
    
    [self setUIElementsHidden: YES];
}

#pragma mark - Action Methods

- (IBAction)loadNativeAd:(id)sender
{
    self.precacheButton.enabled = NO;
    self.showButton.enabled = NO;
    
    self.impressionStatusLabel.text = @"No impression to track";
    
    [[ALSdk shared].nativeAdService loadNextAdAndNotify: self];
}

- (IBAction)precacheNativeAd:(id)sender
{
    // You can use our pre-caching to retrieve assets (app icon, ad image, ad video) locally. OR you can do it with your preferred caching framework.
    // iconURL, imageURL, videoURL needs to be retrieved manually before you can render them
    
    [[ALSdk shared].nativeAdService precacheResourcesForNativeAd: self.nativeAd andNotify: self];
}

- (IBAction)showNativeAd:(id)sender
{
    self.nativeAd = self.cachedNativeAd;
    
    self.appIcon.image = [UIImage imageWithData: [NSData dataWithContentsOfURL: self.nativeAd.iconURL]]; // Local URL
    self.titleLabel.text = self.nativeAd.title;
    self.descriptionLabel.text = self.nativeAd.descriptionText;
    [self.ctaButton setTitle: self.nativeAd.ctaText forState: UIControlStateNormal];
    
    NSString *filename = [NSString stringWithFormat: @"Star_Sprite_%@", self.nativeAd.starRating.stringValue];
    self.rating.image = [UIImage imageNamed: filename];
    
    // NOTE - Videos have aspect ratio of 1:1.85
    [self.mediaView renderViewForNativeAd: self.nativeAd];
    
    [self setUIElementsHidden: NO];
    
    //
    // You are responsible for firing impressions
    //
    [self trackImpression: self.nativeAd];
    
    [self.view layoutIfNeeded];
}

- (IBAction)ctaPressed:(id)sender
{
    [self.nativeAd launchClickTarget];
}

#pragma mark - Impressing Tracking

- (void)trackImpression:(ALNativeAd *)ad
{
    // Callbacks may not happen on main queue
    dispatch_async(dispatch_get_main_queue(), ^{
        self.impressionStatusLabel.text = @"Tracking impression...";
    });
    
    [ad trackImpressionAndNotify: self];
}

- (void)postbackService:(ALPostbackService *)postbackService didExecutePostback:(NSURL *)postbackURL
{
    // Callbacks may not happen on main queue
    dispatch_async(dispatch_get_main_queue(), ^{
        // Impression tracked!
        self.impressionStatusLabel.text = @"Impression tracked";
    });
}

- (void)postbackService:(ALPostbackService *)postbackService didFailToExecutePostback:(nullable NSURL *)postbackURL errorCode:(NSInteger)errorCode
{
    // Callbacks may not happen on main queue
    dispatch_async(dispatch_get_main_queue(), ^{
        // Impression could not be tracked. Retry the postback later.
        self.impressionStatusLabel.text = [NSString stringWithFormat: @"Impression failed to track with error code %ld", errorCode];
    });
}

#pragma mark - Native Ad Load Delegate

- (void)nativeAdService:(ALNativeAdService *)service didLoadAds:(NSArray *)ads
{
    [self logCallback: __PRETTY_FUNCTION__];

    // Callbacks may not happen on main queue
    dispatch_async(dispatch_get_main_queue(), ^{
        self.nativeAd = [ads firstObject];
        self.precacheButton.enabled = YES;
    });
}

- (void)nativeAdService:(ALNativeAdService *)service didFailToLoadAdsWithError:(NSInteger)code
{
    [self logCallback: __PRETTY_FUNCTION__];
}

#pragma mark - Native Ad Precache Delegate

- (void)nativeAdService:(ALNativeAdService *)service didPrecacheImagesForAd:(ALNativeAd *)ad
{
    [self logCallback: __PRETTY_FUNCTION__];
}

- (void)nativeAdService:(ALNativeAdService *)service didPrecacheVideoForAd:(ALNativeAd *)ad
{
    // This delegate method will get called whether an ad actually has a video to precache or not
    [self logCallback: __PRETTY_FUNCTION__];

    // Callbacks may not happen on main queue
    dispatch_async(dispatch_get_main_queue(), ^{
        self.cachedNativeAd = ad;
        self.showButton.enabled = YES;
        self.precacheButton.enabled = NO;
    });
}

- (void)nativeAdService:(ALNativeAdService *)service didFailToPrecacheImagesForAd:(ALNativeAd *)ad withError:(NSInteger)errorCode
{
    [self logCallback: __PRETTY_FUNCTION__];
}

- (void)nativeAdService:(ALNativeAdService *)service didFailToPrecacheVideoForAd:(ALNativeAd *)ad withError:(NSInteger)errorCode
{
    [self logCallback: __PRETTY_FUNCTION__];
}

#pragma mark - Utility

- (void)setUIElementsHidden:(BOOL)hidden
{
    self.appIcon.hidden = hidden;
    self.titleLabel.hidden = hidden;
    self.rating.hidden = hidden;
    self.descriptionLabel.hidden = hidden;
    self.mediaView.hidden = hidden;
    self.ctaButton.hidden = hidden;
}

@end

File: AppLovin MAX Demo App - Swift/AppLovin MAX Demo App - Swift/AppLovin/Native/Programattic/ALDemoNativeAdProgrammaticViewController.swift
//
//  ALDemoNativeAdProgrammaticViewController.swift
//  iOS-SDK-Demo
//
//  Created by Thomas So on 9/25/15.
//  Copyright © 2015 AppLovin. All rights reserved.
//

import UIKit

// Additional documentation - https://applovin.com/integration#iosNative

class ALDemoNativeAdProgrammaticViewController : ALBaseAdViewController
{
    @IBOutlet weak var precacheButton: UIBarButtonItem!
    @IBOutlet weak var showButton: UIBarButtonItem!
    
    @IBOutlet weak var impressionStatusLabel: UILabel!
    
    @IBOutlet weak var appIcon: UIImageView!
    @IBOutlet weak var titleLabel: UILabel!
    @IBOutlet weak var rating: UIImageView!
    @IBOutlet weak var descriptionLabel: UILabel!
    @IBOutlet weak var mediaView: ALCarouselMediaView!
    @IBOutlet weak var ctaButton: UIButton!
    
    var nativeAd: ALNativeAd?
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        appIcon.layer.masksToBounds = true
        appIcon.layer.cornerRadius = 3
        
        ctaButton.layer.masksToBounds = true
        ctaButton.layer.cornerRadius = 3
        
        setUIElementsHidden(true)
    }
    
    // MARK: Action Methods
    
    @IBAction func loadNativeAd(_ sender: AnyObject!)
    {
        logCallback()

        precacheButton.isEnabled = false
        showButton.isEnabled = false
        
        impressionStatusLabel.text = "No impression to track"
        
        ALSdk.shared()!.nativeAdService.loadNextAdAndNotify(self)
    }
    
    @IBAction func precacheNativeAd(_ sender: AnyObject!)
    {
        // You can use our pre-caching to retrieve assets (app icon, ad image, ad video) locally. OR you can do it with your preferred caching framework.
        // iconURL, imageURL, videoURL needs to be retrieved manually before you can render them
        
        logCallback()

        if let ad = nativeAd
        {
            ALSdk.shared()!.nativeAdService.precacheResources(for: ad, andNotify: self)
        }
    }
    
    @IBAction func showNativeAd(_ sender: AnyObject!)
    {
        logCallback()

        if let ad = nativeAd, let iconURL = ad.iconURL
        {
            if let imageData = try? Data(contentsOf: iconURL)
            {
                appIcon.image = UIImage(data: imageData ) // Local URL
            }
            
            titleLabel.text = ad.title
            descriptionLabel.text = ad.descriptionText
            ctaButton.setTitle(ad.ctaText, for: .normal)
            
            let starFilename = "Star_Sprite_\(String(describing: ad.starRating?.stringValue))"
            rating.image = UIImage(named: starFilename)
            
            // NOTE - Videos have aspect ratio of 1:1.85
            mediaView.renderView(for:  ad)
            
            setUIElementsHidden(false)
            
            //
            // You are responsible for firing all necessary postback URLs
            //
            trackImpression(ad)
            
            view.layoutIfNeeded()
        }
    }
    
    @IBAction func ctaPressed(_ sender: AnyObject!)
    {
        nativeAd?.launchClickTarget()
    }
    
    func trackImpression(_ ad: ALNativeAd!)
    {
        // Callbacks may not happen on main queue
        DispatchQueue.main.async {
            ad.trackImpressionAndNotify(self)
        }
    }
    
    func setUIElementsHidden(_ hidden: Bool)
    {
        appIcon.isHidden = hidden
        titleLabel.isHidden = hidden
        rating.isHidden = hidden
        descriptionLabel.isHidden = hidden
        mediaView.isHidden = hidden
        ctaButton.isHidden = hidden
    }
}

extension ALDemoNativeAdProgrammaticViewController : ALPostbackDelegate
{
    func postbackService(_ postbackService: ALPostbackService, didExecutePostback postbackURL: URL)
    {
        // Callbacks may not happen on main queue
        DispatchQueue.main.async {
            // Impression tracked!
            self.impressionStatusLabel.text = "Impression tracked"
        }
    }
    
    func postbackService(_ postbackService: ALPostbackService, didFailToExecutePostback postbackURL: URL?, errorCode: Int)
    {
        // Callbacks may not happen on main queue
        DispatchQueue.main.async {
            // Impression could not be tracked. Retry the postback later.
            self.impressionStatusLabel.text = "Impression failed to track with error code \(errorCode)"
        }
    }
}

extension ALDemoNativeAdProgrammaticViewController : ALNativeAdLoadDelegate
{
    func nativeAdService(_ service: ALNativeAdService, didLoadAds ads: [Any])
    {
        logCallback()

        // Callbacks may not happen on main queue
        DispatchQueue.main.async {
            self.nativeAd = ads.first as? ALNativeAd
            self.precacheButton.isEnabled = true
        }
    }
    
    func nativeAdService(_ service: ALNativeAdService, didFailToLoadAdsWithError code: Int)
    {
        logCallback()
    }
}

extension ALDemoNativeAdProgrammaticViewController : ALNativeAdPrecacheDelegate
{
    func nativeAdService(_ service: ALNativeAdService, didPrecacheImagesFor ad: ALNativeAd)
    {
        logCallback()
    }
    
    func nativeAdService(_ service: ALNativeAdService, didPrecacheVideoFor ad: ALNativeAd)
    {
        // This delegate method will get called whether an ad actually has a video to precache or not
        logCallback()

        // Callbacks may not happen on main queue
        DispatchQueue.main.async {
            self.showButton.isEnabled = true
            self.precacheButton.isEnabled = false
        }
    }
    
    func nativeAdService(_ service: ALNativeAdService, didFailToPrecacheImagesFor ad: ALNativeAd, withError errorCode: Int)
    {
        logCallback()
    }
    
    func nativeAdService(_ service: ALNativeAdService, didFailToPrecacheVideoFor ad: ALNativeAd, withError errorCode: Int)
    {
        logCallback()
    }
}
